### PR TITLE
Attribute, operators, extern modifier fixes

### PR DESCRIPTION
**Pull request by esdrubal**
Fixes 3 issues found while compiling generated code for v4.6.1 mscorlib.dll.

Attributes not visible outside the assembly are now filtered.
Attributes using arguments with typeof expression of a type not visible
outside assembly are also filtered.

When we are generating structs for builtin types such as System.Double, we
cannot use the keyword double while declaring operators.

This fixes bad generated methods such as:
public static decimal operator +(decimal d1, decimal d2) { throw null; }
public static implicit operator decimal (char value) { throw null; }

This changes the generator to not replace a type for its keyword when we
are generating the type itself.

DllImportAttribute was added to methods without modifier extern,
this would cause an error while compiling them.

Fixes:
https://github.com/dotnet/buildtools/issues/869
https://github.com/dotnet/buildtools/issues/904
https://github.com/dotnet/buildtools/issues/658

### CODE REVIEW

File: src/Microsoft.Cci.Extensions/Writers/CSharp/CSDeclarationWriter.Methods.cs [hunk 1]
```diff
@@ -50,6 +50,13 @@ private void WriteDestructor(IMethodDefinition method)
             WriteEmptyBody();
         }
 
+        private void WriteTypeName(ITypeReference type, ITypeReference containingType, bool isDynamic = false)
+        {
+            var useKeywords = containingType.GetTypeName() != type.GetTypeName ();
```
Comment (weshaggard) [inline src/Microsoft.Cci.Extensions/Writers/CSharp/CSDeclarationWriter.Methods.cs]: nit: `type.GetTypeName();` you ad an extra space.

**Comment by weshaggard** [comment]
Thanks for submitting these fixes @esdrubal. The changes look good only one minor nit.

**Comment by esdrubal** [comment]
@weshaggard thanks for looking into this, nit fixed.

**Comment by weshaggard** [comment]
@esdrubal Thanks for the contribution.

**Comment by ericstj** [comment]
Awesome, thanks for this fix!

**Pull request merged**